Implement Question CRUD Operations and Saved Questions Feature

boghtml opened the pull request:
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)


## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions


## Testing Functionality

For testing this functionality, I have provided a link to my Postman collection:

[Postman Collection](https://educational-platform-7691.postman.co/workspace/SoftServe---work~e04aa574-5147-4873-a9d0-3189be5b56d5/collection/37235075-673ce70c-2b91-4f08-8cec-ec264cc0bc29?action=share&creator=37235075)

You can use all the Endpoints in the main directory and utilize the **Saved** folder, as the relevant Endpoints for this PR are currently located there.

Additionally, don't forget to add the `.env` file to the project directory with the following configuration:

```env
# Server
SERVER_PORT=8080

# ScyllaDB
SCYLLA_HOSTS=scylladb:9042
SCYLLA_KEYSPACE=qa_forum
SCYLLA_USERNAME=
SCYLLA_PASSWORD=

boghtml committed:
Implement Question CRUD Operations and Saved Questions Feature

boghtml committed:
removed unnecessary fields in the tables: author_name, profile_url, views_count

VovaD199 commented on BackEnd/forum-service/qa-forum-service/internal/repository/interfaces.go:
Methods like GetAllQuestions, GetQuestionsByAuthor, and GetQuestionsByStatus return entire datasets.
Problem: Inefficient for large datasets, can exhaust memory.
Solution: Introduce pagination:
GetAllQuestions(limit int, pagingState []byte) ([]models.Question, []byte, error)

VovaD199 commented on BackEnd/forum-service/qa-forum-service/internal/repository/interfaces.go:
Batch Operations for Efficiency
Methods like SaveQuestion and UnsaveQuestion operate on single items, which can be slow when handling bulk saves.
Solution: Consider batch processing:
SaveQuestions(userID int, questionIDs []gocql.UUID) error
UnsaveQuestions(userID int, questionIDs []gocql.UUID) error

NikitaPash reviewed (approved):
Except for the moments, pointed out by mentor, it looks nice.

boghtml committed:
fixed logging, fixed the work with SavedQuestion, added pagination with limit\pagination_state, added a stub for authorization

refactor(pagination): remove direct usage of LIMIT in queries & rely on driver-based PageSize

- Replaced "SELECT ... WHERE ... LIMIT ?" with queries that omit LIMIT in CQL.
- Now using PageSize(limit) + PageState(...) in driver (gocql) to handle pagination.
- Updated methods:
  - GetAllQuestions
  - GetQuestionsByAuthor
  - GetQuestionsByStatus
  - GetSavedQuestions
- Ensured they return a non-empty paging_state when the total rows exceed the limit.
- This fix allows multi-page retrieval of large datasets instead of a single 'limited' chunk.

boghtml commented:
Added and fixed everything:
*Added logging, fixed work with SavedQuestion, added pagination with limit\pagination_state, added a stub for authorization

Except for the comment about:
Batch Operations for Efficiency
Methods like SaveQuestion and UnsaveQuestion operate on single items, which can be slow when handling bulk saves.
Solution: Consider batch processing:
SaveQuestions(userID int, questionIDs []gocql.UUID) error
UnsaveQuestions(userID int, questionIDs []gocql.UUID) error

* **the user cannot save multiple questions at the same time, only one by one or what did you mean?**

VovaD199 commented on BackEnd/forum-service/qa-forum-service/internal/handlers/question_handler.go:
```diff
@@ -11,3 +19,249 @@ type Handler struct {
+		return
+	}
+
+	question.Title = updateData.Title
```
The code accepts any JSON data, even if Title or Description are empty.
if question.Title == "" || question.Description == "" {
    c.JSON(http.StatusBadRequest, gin.H{"error": "Title and Description are required"})
    return
}

VovaD199 commented on BackEnd/forum-service/qa-forum-service/internal/handlers/question_handler.go:
```diff
@@ -11,3 +19,249 @@ type Handler struct {
+		return
+	}
+
+	question, err := h.repo.GetQuestionByID(id)
```
If h.repo.GetQuestionByID(id) returns nil, nil, the code does not check for this case.

question, err := h.repo.GetQuestionByID(c.Request.Context(), id)
if err != nil {
	log.Printf("Error fetching question: %v", err)
	c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
	return
}
if question == nil {
	c.JSON(http.StatusNotFound, gin.H{"error": "Question not found"})
	return
}

VovaD199 commented on BackEnd/forum-service/qa-forum-service/internal/handlers/question_handler.go:
```diff
@@ -11,3 +19,249 @@ type Handler struct {
+		return
+	}
+
+	question, err := h.repo.GetQuestionByID(id)
```
If h.repo.GetQuestionByID(id) returns nil, nil, the code does not check for this case.

VovaD199 commented on BackEnd/forum-service/qa-forum-service/internal/repository/question_repository.go:
Error checking in iter.Close()
In functions that work with Iter(), you must first close iter and check for errors before continuing processing.

This problem occurs 6 times in the code

VovaD199 commented on BackEnd/forum-service/qa-forum-service/internal/repository/question_repository.go:
```diff
@@ -0,0 +1,379 @@
+
+	checkQuery := `SELECT user_id FROM saved_questions WHERE user_id = ? AND question_id = ? LIMIT 1`
+	iter := db.session.Query(checkQuery, userID, questionID).Iter()
+	if iter.NumRows() > 0 {
```
NumRows() check not working in Cassandra/Scylla
But NumRows() in Iter() does not give the expected result in ScyllaDB. It is better to use Scan() and check for an error.

VovaD199 commented on BackEnd/forum-service/qa-forum-service/internal/repository/question_repository.go:
```diff
@@ -0,0 +1,379 @@
+	return db.session.Query(query, id).Exec()
+}
+
+func (db *ScyllaDB) SaveQuestion(userID int, questionID gocql.UUID) error {
```
The code to update saves_count is duplicated in SaveQuestion and UnsaveQuestion. It is worth moving it to a helper function.

VovaD199 commented on BackEnd/forum-service/qa-forum-service/internal/repository/question_repository.go:
```diff
@@ -0,0 +1,379 @@
+		return fmt.Errorf("error inserting into questions_by_author table: %w", err)
+	}
+
+	statusQuery := `INSERT INTO questions_by_status (
```
If a question changes status, the entry in questions_by_status is not updated correctly (you need to delete from the old status and add to the new one).
It is worth adding logic to correctly update the status.

P.S. This problem occurs in the file several times.

boghtml committed:
added checks for questions, validation of fields: Title/Description, replaced: iter.NumRows() with iter.Scan(...), moved repeated logic separately

boghtml commented:
Added nil check and Title/Description validation in question_handler.go.
In question_repository.go:
Replaced NumRows() with Scan().
Check iter.Close() before processing data (do this for all methods with Iter()).
Added status update logic in questions_by_status.
Moved saves_count update to helper function.
